drivers/disk/flashdisk.c

```diff
@@ -52,38 +53,91 @@ static int disk_flash_access_status(struct disk_info *disk)
 
 static int disk_flash_access_init(struct disk_info *disk)
 {
-	if (flash_dev) {
-		return 0;
+	struct flashdisk_data *ctx;
+	const struct flash_area *fap = NULL;
+	int rc;
+	struct flash_pages_info page;
+
+	ctx = CONTAINER_OF(disk, struct flashdisk_data, info);
+
+	rc = flash_area_open(ctx->area_id, &fap);
+	if (rc < 0 || fap == NULL) {
+		return -ENODEV;
 	}
 
-	flash_dev = device_get_binding(CONFIG_DISK_FLASH_DEV_NAME);
-	if (!flash_dev) {
+	disk->dev = flash_area_get_device(fap);
+	if (disk->dev == NULL) {
+		flash_area_close(fap);
 		return -ENODEV;
 	}
 
+	rc = flash_get_page_info_by_offs(disk->dev, ctx->offset, &page);
+	if (rc != 0) {
+		LOG_ERR("Error %d while getting page info", rc);
+		flash_area_close(fap);
+		return rc;
+	}
+
+	ctx->write_bs = flash_get_write_block_size(disk->dev);
+	ctx->page_size = page.size;
+	LOG_INF("Initialize device %s", disk->name);
+	LOG_INF("offset %lx, write bs %zu, page size %zu, volume size %zu",
+		(long)ctx->offset, ctx->write_bs, ctx->page_size, ctx->size);
+
+	flash_area_close(fap);
```

Comment:
The driver currently extracts the flash area device and information, and then uses the device directly. I think flashdisk could actually use the flash area API to access the device. If flashdisk would use the flash area API for read/write/erase then the bounds check would be done by `is_in_flash_area_bounds()` and wouldn't require custom checks in flashdisk.

Reply:
> The driver currently extracts the flash area device and information, and then uses the device directly. I think flashdisk could actually use the flash area API to access the device. If flashdisk would use the flash area API for read/write/erase then the bounds check would be done by `is_in_flash_area_bounds()` and wouldn't require custom checks in flashdisk.

Yes this is true, you could do that.  I have not mentioned that, because I am not insisting on ding so.
Why I think the flash disk should keep flash area open is that it should keep its claim over the area.
I know that the `flash_area_open` does no checks, currently, whether area is opened or in use already.

Reply:
I will simply remove the `flash_area_close()` call (it is no-op anyway) for this PR. The driver uses flash directly and there is no "uninit" function so once initialized, the flash area will be claimed indefinitely.

Flashdisk driver does assume homogeneous page size through the whole flash area. Removing that assumption is probably best to be done together with caching (if at all).